kornia/geometry/vector.py

```diff
@@ -1,79 +1,81 @@
     def __init__(self, vector: Tensor) -> None:
-        super().__init__()
-        # KORNIA_CHECK_SHAPE(vector, ["B", "N"])  # FIXME: resolve shape bugs. @edgarriba
-        self._vector = Parameter(vector)
+        super().__init__(vector)
+        KORNIA_CHECK(vector.shape[-1] == 3)
 
     @property
-    def data(self) -> Tensor:
-        return self._vector
+    def x(self) -> Tensor:
+        return self.data[..., 0]
 
     @property
-    def device(self):
-        return self.data.device
+    def y(self) -> Tensor:
+        return self.data[..., 1]
 
     @property
-    def dtype(self):
-        return self.data.dtype
+    def z(self) -> Tensor:
+        return self.data[..., 2]
 
-    @property
-    def shape(self):
-        return self.data.shape
-
-    def __repr__(self) -> str:
-        return f"{self.data}"
-
-    @classmethod
-    def random(cls, shape: List[int], device=None, dtype=None) -> "_VectorType":
-        return cls(rand(shape, device=device, dtype=dtype))
-
-    def normalized(self) -> "_VectorType":
-        vec_norm = normalize(self._vector, p=2, dim=-1)
-        return _VectorType(vec_norm)
+    def normalized(self) -> "Vector3":
+        return Vector3(normalize(self.data, p=2, dim=-1))
 
-    def squared_norm(self) -> "_VectorType":
-        return _VectorType(_squared_norm(self.data))
+    def dot(self, right: "Vector3") -> Scalar:
+        return Scalar(batched_dot_product(self.data, right.data))
 
-    def dot(self, right: "_VectorType") -> "_VectorType":
-        if len(right.shape) == 1:
-            return _VectorType(self.data.dot(right.data))
-        return _VectorType(batched_dot_product(self.data, right.data))
-
-
-class UnitVector(Module):
-    def __init__(self, vector: Union[Tensor, _VectorType]) -> None:
-        super().__init__()
-        KORNIA_CHECK_SHAPE(vector, ["B", "N"])
-        if isinstance(vector, Tensor):
-            self._vector = Parameter(vector)
-
-    @property
-    def vector(self) -> Tensor:
-        return self._vector
+    def squared_norm(self) -> Scalar:
+        return Scalar(batched_squared_norm(self.data))
 
     @classmethod
-    def from_unit_vector(cls, v: Tensor) -> "UnitVector":
-        # TODO: add checks https://github.com/strasdat/Sophus/blob/23.04-beta/cpp/sophus/geometry/ray.h#L59
-        return UnitVector(_VectorType(v))
+    def random(cls, shape: Optional[Tuple[int, ...]] = None, device=None, dtype=None) -> "Vector3":
+        if shape is None:
+            shape = ()
+        return cls(rand(shape + (3,), device=device, dtype=dtype))
 
     @classmethod
-    def from_vector(cls, v: Tensor) -> "UnitVector":
-        """From a vector and normalize."""
-        return UnitVector(_VectorType(v).normalized())
+    def from_coords(
+        cls, x: Union[float, Tensor], y: Union[float, Tensor], z: Union[float, Tensor], device=None, dtype=None
+    ) -> "Vector3":
+        if not (isinstance(x, Tensor) and isinstance(y, Tensor) and isinstance(z, Tensor)):
```

Comment:
If just one input is a float and others are tensor, i think this will fail

Reply:
maybe the typing can be

```python
    @overload
    def from_coords(
        cls, x: Tensor, y: Tensor, z: Tensor, device: Device=None, dtype: torch.dtype=None
    ) -> "Vector3":
    
    @overload
    def from_coords(
        cls, x: float, y: float, z: float, device: Device=None, dtype: torch.dtype=None
    ) -> "Vector3":
    
    def from_coords(...):
```
this way mypy can help

Reply:
not clear to me a good pattern here. look at the solution i provided ... very ugly though